source/common/listener_manager/active_tcp_socket.cc

```diff
@@ -128,30 +128,33 @@ void ActiveTcpSocket::continueFilterChain(bool success) {
           no_error = false;
           break;
         } else {
-          // If the listener maxReadBytes() is 0, then it shouldn't return
-          // `FilterStatus::StopIteration` from `onAccept` to wait for more data.
-          ASSERT((*iter_)->maxReadBytes() != 0);
-          if (listener_filter_buffer_ == nullptr) {
-            if ((*iter_)->maxReadBytes() > 0) {
+          // There are two cases for returning StopIteration. One is listener filter
+          // needs to wait for data. The other is listener filter with 0 maxReadBytes()
+          // doesn't inspect data, but needs to wait for some asynchronous callback.
+          if ((*iter_)->maxReadBytes() > 0) {
+            if (listener_filter_buffer_ == nullptr) {
               createListenerFilterBuffer();
+            } else {
+              // If the current filter expect more data than previous filters, then
+              // increase the filter buffer's capacity.
+              if (listener_filter_buffer_->capacity() < (*iter_)->maxReadBytes()) {
+                listener_filter_buffer_->resetCapacity((*iter_)->maxReadBytes());
+              }
             }
-          } else {
-            // If the current filter expect more data than previous filters, then
-            // increase the filter buffer's capacity.
-            if (listener_filter_buffer_->capacity() < (*iter_)->maxReadBytes()) {
-              listener_filter_buffer_->resetCapacity((*iter_)->maxReadBytes());
+            if (listener_filter_buffer_ != nullptr) {
+              // There are two cases for activate event manually: One is
+              // the data is already available when connect, activate the read event to peek
+              // data from the socket . Another one is the data already
+              // peeked into the buffer when previous filter processing the data, then activate the
+              // read event to trigger the current filter callback to process the data.
+              listener_filter_buffer_->activateFileEvent(Event::FileReadyType::Read);
             }
+            // Waiting for more data.
+            return;
+          } else {
+            // Waiting for asynchronous callback.
+            return;
```

Comment:
you still can crash at here https://github.com/envoyproxy/envoy/blob/51e253405a2be7f94df8c0ba78bd884dc79bb8a5/source/common/listener_manager/active_tcp_socket.cc#L86, if the previous listener filter expect more data, then the listener filter buffer will be created and begin to listen on the data.

Reply:
That's expected, only listener filter with maxReadBytes() > 0 interacts with listener filter buffer. For listener filter with maxReadBytes() = 0, only onAccept is called

Reply:
If you have multiple listener filter, the first listener filter with maxReadBytes() > 0, the second listener filter with maxReadBytes() == 0. It means the ListenerFilterBuffer is created for the first listener filter, and it will continue to listen the read event. And there may have a chance, new data is coming when iterate the second listener filter.

Reply:
I see your point, and I'll move createListenerFilterBuffer out so that it'll be called the first time StopIteration is returned by onAccept. And with this change, if the first listener filter has maxReadBytes() = 0, it will create a buffer with length of 0, which may cause issues for non-blocking recv call, so it looks to me I need to handle this corner case in ListenerFilterBuffer, could you please share some of your thoughts?

Reply:
I think we can handle it in ListenreFilterBuffer's onData callback

Reply:
Could you please elaborate on the change? My concern is on peekFromSocket before onData callback
https://github.com/envoyproxy/envoy/blob/1a8088436550bc6433d2c411c573189ea9c080c9/source/common/network/listener_filter_buffer_impl.cc#L55-L59

Reply:
One way is you can set a 1 byte buffer size for the ListenerFilterBuffer even the filter's maxReadBytes is 0. Or you can handle more complex, like if maxReadBytes is 0, then you create ListenerFilterBuffer with 0 size buffer, then you only listen on the Close event (I wasn't sure the remote close will trigger the close event, you need to give a test, ensure it works as expected).

Reply:
I tried both and finally decided to go with setting buffer size to 1, more details in latest conversation